istioctl/cmd/revision.go

```diff
@@ -0,0 +1,326 @@
+			revToIOP[revision] = []*iopv1alpha1.IstioOperator{iop}
+		} else {
+			revToIOP[revision] = append(revToIOP[revision], iop)
+		}
+	}
+	w := new(tabwriter.Writer).Init(writer, 0, 8, 1, ' ', 0)
+	if verbose {
+		_, _ = fmt.Fprintln(w, "REVISION\tIOP\tPROFILE\tCOMPONENTS\tCUSTOMIZATIONS")
+	} else {
+		_, _ = fmt.Fprintln(w, "REVISION\tIOP\tPROFILE\tCOMPONENTS")
+	}
+	for revision, iopWithRev := range revToIOP {
+		for i, iop := range iopWithRev {
+			iops := iop.Spec
+			components := getEnabledComponents(iops)
+			customizations := []string{}
+			if verbose {
+				var err error
+				customizations, err = getDiffs(iop, manifestsPath, effectiveProfile(iop.Spec.Profile), l)
+				if err != nil {
+					l.LogAndErrorf("error while computing customization: %s", err.Error())
+				}
+			}
+			// The first row is for printing revision, IOP related information
+			numRows := max(len(components), max(1, len(customizations)))
+			for j := 0; j < numRows; j++ {
+				var outRevision string
+				if i == 0 && j == 0 {
+					outRevision = renderWithDefault(revision, "<default>")
+				}
+				var outIopName, outProfile string
+				if j == 0 {
+					outIopName = fmt.Sprintf("%s/%s", iop.Namespace, iop.Name)
+					outProfile = renderWithDefault(iops.GetProfile(), "default")
+				}
+				outCompName := ""
+				if j < len(components) {
+					outCompName = components[j]
+				}
+				outCustomization := ""
+				if j < len(customizations) {
+					outCustomization = customizations[j]
+				}
+				if verbose {
+					fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\n",
+						outRevision, outIopName, outProfile, outCompName, outCustomization)
+				} else {
+					fmt.Fprintf(w, "%s\t%s\t%s\t%s\n",
+						outRevision, outIopName, outProfile, outCompName)
+				}
+			}
+		}
+	}
+	return w.Flush()
+}
+
+func getEnabledComponents(iops *v1alpha1.IstioOperatorSpec) []string {
+	enabledComponents := []string{}
+	if iops.Components.Base.Enabled.GetValue() {
+		enabledComponents = append(enabledComponents, "base")
+	}
+	if iops.Components.Cni.Enabled.GetValue() {
+		enabledComponents = append(enabledComponents, "cni")
+	}
+	if iops.Components.Pilot.Enabled.GetValue() {
+		enabledComponents = append(enabledComponents, "istiod")
+	}
+	if iops.Components.IstiodRemote.Enabled.GetValue() {
+		enabledComponents = append(enabledComponents, "istiod-remote")
+	}
+	for _, gw := range iops.Components.IngressGateways {
+		if gw.Enabled.Value {
+			enabledComponents = append(enabledComponents, fmt.Sprintf("ingress:%s", gw.GetName()))
+		}
+	}
+	for _, gw := range iops.Components.EgressGateways {
+		if gw.Enabled.Value {
+			enabledComponents = append(enabledComponents, fmt.Sprintf("egress:%s", gw.GetName()))
+		}
+	}
+	return enabledComponents
+}
+
+func getIOPs(restConfig *rest.Config) ([]*iopv1alpha1.IstioOperator, error) {
+	client, err := dynamic.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	ul, err := client.
+		Resource(istioOperatorGVR).
+		Namespace(istioNamespace).
```

Comment:
Will this work if the Istio is installed in custom namespace?

Reply:
Yeah. You have to specify with `-i` flag. For instance, I ran this command
```
$ istioctl x revision list -i kube-system
No IstioOperators present.
```

Reply:
i think that's good enough, as long as all user namespaces are scanned for injected pods (you can probably exclude kube-system and kube-public).